File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    const stack = [-1]
    let res = 0
    for (let i=0; i < s.length; i++) {
        const c = s[i]
        if (c === "(") {
            stack.push(i)
        } else {
            stack.pop()
            if (!stack.length) {
                stack.push(i)
            } else {
                res = Math.max(res, i - stack[stack.length-1])
            }
        }
    }
    return res
};

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack, res = [-1], 0
        for i, c in enumerate(s):
            if c == "(":
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    res = max(res, i - stack[-1])
        return res
            